File: backend/main/java/de/griefed/serverpackcreator/spring/serverpack/ServerPackService.java
/* Copyright (C) 2022  Griefed
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301
 * USA
 *
 * The full license can be found at https:github.com/Griefed/ServerPackCreator/blob/main/LICENSE
 */
package de.griefed.serverpackcreator.spring.serverpack;

import java.net.MalformedURLException;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.sql.Timestamp;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.core.io.Resource;
import org.springframework.core.io.UrlResource;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;

/**
 * Class revolving around with server packs, like downloading, retrieving, deleting, voting etc.
 *
 * @author Griefed
 */
@Service
public class ServerPackService {

  private static final Logger LOG = LogManager.getLogger(ServerPackService.class);

  private final ServerPackRepository SERVERPACKREPOSITORY;

  /**
   * Constructor responsible for our DI.
   *
   * @param injectedServerPackRepository Instance of {@link ServerPackRepository}.
   * @author Griefed
   */
  @Autowired
  public ServerPackService(ServerPackRepository injectedServerPackRepository) {
    this.SERVERPACKREPOSITORY = injectedServerPackRepository;
  }

  /**
   * Download a server pack with the given database id.
   *
   * @param id Integer. The database id of the server pack to download.
   * @return Returns a curseResponse entity with either the server pack as a downloadable file, or a
   *     curseResponse entity with a not found body.
   * @author Griefed
   */
  public ResponseEntity<Resource> downloadServerPackById(int id) {
    if (SERVERPACKREPOSITORY.findById(id).isPresent()
        && SERVERPACKREPOSITORY.findById(id).get().getStatus().matches("Available")) {

      ServerPackModel serverPackModel = SERVERPACKREPOSITORY.findById(id).get();

      Path path = Paths.get(serverPackModel.getPath());
      Resource resource = null;
      String contentType = "application/zip";

      try {
        resource = new UrlResource(path.toUri());
      } catch (MalformedURLException ex) {
        LOG.error("Error generating download for server pack with ID" + id + ".", ex);
      }

      updateDownloadCounter(id);

      return ResponseEntity.ok()
          .contentType(MediaType.parseMediaType(contentType))
          .header(
              HttpHeaders.CONTENT_DISPOSITION,
              "attachment; filename=\""
                  + serverPackModel.getFileDiskName().replace(".zip", "")
                  + "_server_pack.zip"
                  + "\"")
          .body(resource);
    } else {

      return ResponseEntity.notFound().build();
    }
  }

  /**
   * Either upvote or downvote a given server pack.
   *
   * @param voting String. The database id of the server pack and whether it should be up- or
   *     downvoted.
   * @return Returns ok if the vote went through, bad request if the passed vote was malformed, or
   *     not found if the server pack could not be found.
   * @author Griefed
   */
  public ResponseEntity<Object> voteForServerPack(String voting) {
    String[] vote = voting.split(",");
    int id = Integer.parseInt(vote[0]);
    if (SERVERPACKREPOSITORY.findById(id).isPresent()
        && SERVERPACKREPOSITORY.findById(id).get().getStatus().equals("Available")) {

      if (vote[1].equalsIgnoreCase("up")) {

        updateConfirmedCounter(id, +1);
        return ResponseEntity.ok().build();

      } else if (vote[1].equalsIgnoreCase("down")) {

        updateConfirmedCounter(id, -1);
        return ResponseEntity.ok().build();

      } else {

        return ResponseEntity.badRequest().build();
      }

    } else {

      return ResponseEntity.notFound().build();
    }
  }

  /**
   * Get a list of all available server packs.
   *
   * @return List ServerPackModel. Returns a list of all available server packs.
   * @author Griefed
   */
  public List<ServerPackModel> getServerPacks() {
    List<ServerPackModel> serverPackModels = new ArrayList<>();
    SERVERPACKREPOSITORY.findAll().forEach(serverPackModels::add);
    return serverPackModels;
  }

  /**
   * Store a server pack in the database.
   *
   * @param serverPackModel Instance of {@link ServerPackModel} to store in the database.
   * @author Griefed
   */
  public void insert(ServerPackModel serverPackModel) {
    SERVERPACKREPOSITORY.save(serverPackModel);
  }

  /**
   * Update a server pack database entry with the given database id.
   *
   * @param id Integer. The database id of the server pack to initialize.
   * @param serverPackModel Instance of {@link ServerPackModel} with which to initialize the entry
   *     in the database.
   * @author Griefed
   */
  public void updateServerPackByID(int id, ServerPackModel serverPackModel) {
    if (SERVERPACKREPOSITORY.findById(id).isPresent()) {
      ServerPackModel serverPackModelFromDB = SERVERPACKREPOSITORY.findById(id).get();
      LOG.debug("Updating database with: " + serverPackModel.repositoryToString());
      serverPackModelFromDB.setProjectName(serverPackModel.getProjectName());
      serverPackModelFromDB.setFileName(serverPackModel.getFileName());
      serverPackModelFromDB.setFileDiskName(serverPackModel.getFileDiskName());
      serverPackModelFromDB.setSize(serverPackModel.getSize());
      serverPackModelFromDB.setDownloads(serverPackModel.getDownloads());
      serverPackModelFromDB.setConfirmedWorking(serverPackModel.getConfirmedWorking());
      serverPackModelFromDB.setStatus(serverPackModel.getStatus());
      serverPackModelFromDB.setLastModified(new Timestamp(new Date().getTime()));
      serverPackModelFromDB.setPath(serverPackModel.getPath());
      SERVERPACKREPOSITORY.save(serverPackModelFromDB);
    }
  }

  /**
   * Increment the download counter for a given server pack entry in the database identified by the
   * database id.
   *
   * @param id Integer. The database id of the server pack.
   * @author Griefed
   */
  public void updateDownloadCounter(int id) {
    if (SERVERPACKREPOSITORY.findById(id).isPresent()) {
      ServerPackModel serverPackModelFromDB = SERVERPACKREPOSITORY.findById(id).get();
      serverPackModelFromDB.setDownloads(serverPackModelFromDB.getDownloads() + 1);
      SERVERPACKREPOSITORY.save(serverPackModelFromDB);
    }
  }

  /**
   * Either increment or decrement the confirmed working value of a given server pack entry in the
   * database, identified by the database id.
   *
   * @param id Integer. The database id of the server pack.
   * @param vote Integer. Positive for upvote, negative for downvote
   * @author Griefed
   */
  public void updateConfirmedCounter(int id, int vote) {
    if (SERVERPACKREPOSITORY.findById(id).isPresent()) {
      ServerPackModel serverPackModelFromDB = SERVERPACKREPOSITORY.findById(id).get();
      serverPackModelFromDB.setConfirmedWorking(serverPackModelFromDB.getConfirmedWorking() + vote);
      SERVERPACKREPOSITORY.save(serverPackModelFromDB);
    }
  }

  /**
   * Deletes a server pack with the given id.
   *
   * @param id Integer. The database id of the server pack to delete.
   * @author Griefed
   */
  public void deleteServerPack(int id) {
    SERVERPACKREPOSITORY.deleteById(id);
  }
}

File: serverpackcreator-app/src/main/kotlin/de/griefed/serverpackcreator/app/web/serverpack/ServerPackService.kt
/* Copyright (C) 2025 Griefed
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301
 * USA
 *
 * The full license can be found at https:github.com/Griefed/ServerPackCreator/blob/main/LICENSE
 */
package de.griefed.serverpackcreator.app.web.serverpack

import de.griefed.serverpackcreator.api.ApiProperties
import de.griefed.serverpackcreator.app.web.storage.SavedFile
import de.griefed.serverpackcreator.app.web.storage.StorageSystem
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Sort
import org.springframework.data.mongodb.gridfs.GridFsOperations
import org.springframework.data.mongodb.gridfs.GridFsTemplate
import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Service
import java.io.File
import java.nio.file.Path
import java.security.MessageDigest
import java.util.*

/**
 * Class revolving around with server packs, like downloading, retrieving, deleting, voting etc.
 *
 * @author Griefed
 */
@Service
class ServerPackService @Autowired constructor(
    private val serverPackRepository: ServerPackRepository,
    private val serverPackDownloadRepository: ServerPackDownloadRepository,
    gridFsTemplate: GridFsTemplate,
    gridFsOperations: GridFsOperations,
    messageDigestInstance: MessageDigest,
    apiProperties: ApiProperties
) {
    private val rootLocation: Path = apiProperties.serverPacksDirectory.toPath()
    private val storage: StorageSystem = StorageSystem(
        rootLocation,
        messageDigestInstance,
        gridFsTemplate,
        gridFsOperations
    )

    fun getServerPack(id: String): Optional<ServerPack> {
        return serverPackRepository.findById(id)
    }

    /**
     * Increment the download counter for a given server pack entry in the database identified by the
     * database id.
     *
     * @param id The database id of the server pack.
     * @author Griefed
     */
    fun updateDownloadStats(id: String): Optional<ServerPack> {
        val request = serverPackRepository.findById(id)
        if (request.isPresent) {
            val pack = request.get()
            pack.downloads++
            serverPackDownloadRepository.save(ServerPackDownload(pack))
            return Optional.of(serverPackRepository.save(pack))
        } else {
            return Optional.empty()
        }
    }

    /**
     * Increment the download counter for a given server pack entry in the database identified by the
     * database id.
     *
     * @param serverPack The server pack for which to update the download counter.
     * @author Griefed
     */
    @Suppress("unused")
    fun updateDownloadStats(serverPack: ServerPack) {
        serverPack.downloads++
        serverPackRepository.save(serverPack)
    }

    /**
     * Either upvote or downvote a given server pack.
     *
     * @param id The database id of the server pack.
     * @param vote Either "up" or "down".
     * @return Returns ok if the vote went through, bad request if the passed vote was malformed, or
     * not found if the server pack could not be found.
     * @author Griefed
     */
    fun voteForServerPack(id: String, vote: String): ResponseEntity<Any> {
        val pack = serverPackRepository.findById(id)
        return if (pack.isPresent) {
            if (vote.equals("up", ignoreCase = true)) {
                pack.get().confirmedWorking++
                serverPackRepository.save(pack.get())
                ResponseEntity.ok().build()
            } else if (vote.equals("down", ignoreCase = true)) {
                pack.get().confirmedWorking--
                serverPackRepository.save(pack.get())
                ResponseEntity.ok().build()
            } else {
                ResponseEntity.badRequest().build()
            }
        } else {
            ResponseEntity.notFound().build()
        }
    }

    fun getServerPacks(sort: Sort = Sort.by(Sort.Direction.DESC, "dateCreated")): List<ServerPack> {
        return serverPackRepository.findAll(sort)
    }

    fun getServerPacks(
        sizedPage: PageRequest,
        sort: Sort = Sort.by(Sort.Direction.DESC, "dateCreated")
    ): Page<ServerPack> {
        return serverPackRepository.findAll(sizedPage.withSort(sort))
    }

    /**
     * Save a server pack to the database.
     *
     * @author Griefed
     */
    fun saveServerPack(serverPack: ServerPack) {
        serverPackRepository.save(serverPack)
    }

    /**
     * Move the specified file to a new one which will receive a new filename based on the current time in milliseconds.
     * This is done to prevent clashes in case a server pack is generated from an existing modpack, using a new
     * [de.griefed.serverpackcreator.app.web.serverpack.customizing.RunConfiguration].
     *
     * @author Griefed
     */
    fun storeServerPackFile(file: File): SavedFile {
        return storage.store(file).get()
    }

    /**
     * Deletes a server pack with the given id.
     *
     * @param serverPack The serverpack to delete.
     * @author Griefed
     */
    @Suppress("unused")
    fun deleteServerPack(serverPack: ServerPack) {
        serverPackRepository.deleteById(serverPack.id!!)
    }

    @Suppress("unused")
    fun deleteServerPack(id: String) {
        val serverPack = serverPackRepository.findById(id)
        if (serverPack.isPresent) {
            serverPackRepository.deleteById(id)
            storage.delete(serverPack.get().fileID!!)
        }
    }

    /**
     * Get the ZIP-archive of a server pack.
     *
     * @param serverPack The serverpack to be stored to disk from the database
     * @return The modpack-file, wrapped in an [Optional]
     * @author Griefed
     */
    fun getServerPackArchive(serverPack: ServerPack): Optional<File> {
        return storage.load(serverPack.fileID!!)
    }

    @Suppress("unused")
    fun getServerPackView(id: String): Optional<ServerPack> {
        return serverPackRepository.findById(id)
    }
}
